sumdbaudit/audit/service.go

```diff
@@ -0,0 +1,361 @@
+	}
+
+	root, err := logRange.GetRootHash(nil)
+	if err != nil {
+		return fmt.Errorf("failed to get root hash: %v", err)
+	}
+	var rootHash tlog.Hash
+	copy(rootHash[:], root)
+	if rootHash != checkpoint.Hash {
+		return fmt.Errorf("log root mismatch at tree size %d; calculated %x, SumDB says %x", checkpoint.N, root, checkpoint.Hash[:])
+	}
+	return nil
+}
+
+// VerifyTiles checks that every tile calculated locally matches the result returned
+// by SumDB. This shouldn't be possible if CheckRootHash is working, but this may be
+// useful to determine where any corruption has happened in the tree.
+func (s *Service) VerifyTiles(ctx context.Context, checkpoint *tlog.Tree) error {
+	for level := 0; level <= s.getLevelsForLeafCount(checkpoint.N); level++ {
+		finishedLevel := false
+		offset := 0
+		for !finishedLevel {
+			localHashes, err := s.localDB.GetTile(s.height, level, offset)
+			if err != nil {
+				if err == sql.ErrNoRows {
+					finishedLevel = true
+					continue
+				}
+				return fmt.Errorf("failed to get tile hashes: %v", err)
+			}
+			sumDBHashes, err := s.sumDB.TileHashes(level, offset)
+			if err != nil {
+				return fmt.Errorf("failed to get tile hashes: %v", err)
+			}
+
+			for i := 0; i < 1<<s.height; i++ {
+				var lHash tlog.Hash
+				copy(lHash[:], localHashes[i])
+				if sumDBHashes[i] != lHash {
+					return fmt.Errorf("found mismatched hash at L=%d, O=%d, leaf=%d\n\tlocal : %x\n\tremote: %x", level, offset, i, sumDBHashes[i][:], localHashes[i])
+				}
+			}
+			offset++
+		}
+	}
+	return nil
+}
+
+// ProcessMetadata parses the leaf data and writes the semantic data into the DB.
+func (s *Service) ProcessMetadata(ctx context.Context, checkpoint *tlog.Tree) error {
+	tileWidth := 1 << s.height
+	metadata := make([]Metadata, tileWidth)
+	// TODO: skip to head of metadata
+	for offset := 0; offset < int(checkpoint.N/int64(tileWidth)); offset++ {
+		leafOffset := int64(offset) * int64(tileWidth)
+		hashes, err := s.localDB.GetLeaves(leafOffset, tileWidth)
+		if err != nil {
+			return err
+		}
+		for i, h := range hashes {
+			leafID := leafOffset + int64(i)
+
+			lines := strings.Split(string(h), "\n")
+			tokens := strings.Split(lines[0], " ")
+			module, version, repoHash := tokens[0], tokens[1], tokens[2]
+			tokens = strings.Split(lines[1], " ")
+			if got, want := tokens[0], module; got != want {
+				return fmt.Errorf("mismatched module names at %d: (%s, %s)", leafID, got, want)
+			}
+			if got, want := tokens[1][:len(version)], version; got != want {
+				return fmt.Errorf("mismatched version names at %d: (%s, %s)", leafID, got, want)
+			}
+			modHash := tokens[2]
+
+			metadata[i] = Metadata{module, version, repoHash, modHash}
+		}
+		if err := s.localDB.SetLeafMetadata(ctx, leafOffset, metadata); err != nil {
+			return err
+		}
+	}
+	return nil
+}
+
+func (s *Service) hashLeafLevel(tileCount int, roots chan *compact.Range) error {
```

Comment:
nit: probably should define the channel as `roots chan<- *compactRange` if you're only writing to it.

(same elsewhere)